### PR TITLE
Maybe trigger backup at end of sync

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/2938

There might be other way to do that. On some other client there is signaling when a new key is imported to trigger that.
But doing it after the sync is probably reasonable.

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk/src/sliding_sync/client.rs
Comment (BillCarsonFr) [inline crates/matrix-sdk/src/sliding_sync/client.rs:60]: I guess there is no need to check if it's enabled or not. It would be no op if not

**Comment by BillCarsonFr** [comment]
As a side effect it will fix this https://github.com/element-hq/element-x-ios/issues/2236

Because now that key upload works, the sdk will detect that the cache is out of sync and will disable backup.

**Comment by Hywan** [comment]
I think @bnjbvr is the expert here. Re-assigning the PR to him.

File: crates/matrix-sdk/src/sliding_sync/client.rs
Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/client.rs:60]: We need to check encryption is needed. We never know if it’s going to be a no-op in the future. Also we probably need to check some feature flags too.

Comment (BillCarsonFr) [reply crates/matrix-sdk/src/sliding_sync/client.rs:60]: The enclosing method has `#[cfg(feature = "e2e-encryption")]`

Comment (Hywan) [reply crates/matrix-sdk/src/sliding_sync/client.rs:60]: Ah true, sorry I've missed that. Well, it should be good to me then. Waiting on another review from @bnjbvr though.

Comment (bnjbvr) [reply crates/matrix-sdk/src/sliding_sync/client.rs:60]: This will only work for sliding sync. How do you feel about putting this under `preprocess_to_device_events`, which is called by both sliding sync and sync v2?

**Review by bnjbvr — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk/src/sliding_sync/client.rs
Comment (bnjbvr) [reply crates/matrix-sdk/src/sliding_sync/client.rs:60]: @BillCarsonFr let me know that `maybe_trigger_backup()` lives in the high-level SDK crate, while I was proposing to put it in the lower-level sdk-base crate. It's a fair point, so instead can you duplicate this line in the other code path for sync v2 please? (or if you use another mechanism like subscriptions, please re-assign me for review)

Comment (BillCarsonFr) [reply crates/matrix-sdk/src/sliding_sync/client.rs:60]: > @BillCarsonFr let me know that `maybe_trigger_backup()` lives in the high-level SDK crate, while I was proposing to put it in the lower-level sdk-base crate. It's a fair point, so instead can you duplicate this line in the other code path for sync v2 please? (or if you use another mechanism like subscriptions, please re-assign me for review)

Done, added test for both syncs type

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/encryption/backups.rs:680]: nit: indent is a bit broken here

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/encryption/backups.rs:737]: nit: can you inline `invite` and `request` into the `create_room` call? (fewer variables => lower cognitive complexity)

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/encryption/backups.rs:749]: Bad copy pasta, no Peter around here :)

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/encryption/backups.rs:559]: nit: indent seems wrong here

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/encryption/backups.rs:524]: nit: indent seems wrong here

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (bnjbvr) [inline crates/matrix-sdk/tests/integration/encryption/backups.rs:79]: Instead of this function, I'd be fine immediately duplicating this function's body at all the call sites, so we immediately see what's being mocked and how.

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Thanks for writing tests! Small comments about the tests, then this is good to go :+1:

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk/tests/integration/encryption/backups.rs]: done

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk/tests/integration/encryption/backups.rs:721]: I tried, but don't see how to make it work. Looks that there is a problem with the assign! macro. FWIW I just copy pasted this code from other tests

**Review by bnjbvr — APPROVED** [review]
Thanks! (Will address the last comment myself.)

File: crates/matrix-sdk/tests/integration/encryption/backups.rs
Comment (Hywan) [inline crates/matrix-sdk/tests/integration/encryption/backups.rs:535]: It looks like a copy-pasta error: are you sure it's 404? 

@bnjbvr is fixing it in https://github.com/matrix-org/matrix-rust-sdk/pull/3953.

**Pull request merged**
